editoast/editoast_derive/src/model.rs

```diff
@@ -1,55 +1,231 @@
+        enum #errors_name {
+            #[error(#not_found_error)]
+            #[editoast_error(status = 404)]
+            NotFound { obj_id: i64 },
+        }
+    }
+}
+
+fn create_functions(config: &Config) -> TokenStream {
+    // If `create` is not enabled, we don't generate anything
+    if !config.create_enabled {
+        return quote! {};
+    }
+    let table = &config.table;
+    let model_name = &config.model_name;
+    let documentation = format!(
+        r#"
+        Create a new `{model_name}` in the database.
+        Returns the created object with its default values filled (like the id).
+
+        ### Example
+
+        ```
+        let obj = {model_name} {{ id: None, ... }};
+        let created_obj = obj.create(db_pool).await?;
+        let obj_id = created_obj.id.unwrap();
+        ```
+        "#
+    );
+
+    quote! {
+        /// Same as [create](Self::create) but takes a single postgres connection.
+        /// Useful when you are in a transaction.
+        pub fn create_conn(self, conn: &mut diesel::PgConnection) -> crate::error::Result<Self> {
+            use #table::table;
+            use diesel::RunQueryDsl;
+
+            Ok(diesel::insert_into(table).values(&self).get_result(conn)?)
+        }
+
+        #[doc = #documentation]
+        pub async fn create(self, db_pool: actix_web::web::Data<crate::DbPool>) -> crate::error::Result<Self> {
+            use actix_web::web::block;
+
+            block::<_, crate::error::Result<_>>(move || {
+                let mut conn = db_pool.get().expect("Failed to get DB connection");
+                self.create_conn(&mut conn)
+            })
+            .await
+            .unwrap()
+        }
+    }
+}
 
-    proc_macro::TokenStream::from(expanded)
+fn retrieve_functions(config: &Config) -> TokenStream {
+    // If `retrieve` is not enabled, we don't generate anything
+    if !config.retrieve_enabled {
+        return quote! {};
+    }
+    let table = &config.table;
+    let errors_name = &config.errors_name;
+
+    let model_name = &config.model_name;
+    let documentation = format!(
+        r#"
+        Retrieve a `{model_name}` given its ID (primary key).
+        Returns an error if is not found.
+
+        ### Example
+
+        ```
+        let obj = {model_name}::retrieve(db_pool, 42).await?;
+        ```
+        "#
+    );
+
+    quote! {
+        /// Same as [retrieve](Self::retrieve) but takes a single postgres connection.
+        /// Useful when you are in a transaction.
+        pub fn retrieve_conn(conn: &mut diesel::PgConnection, obj_id: i64) -> crate::error::Result<Self> {
+            use #table::table;
+            use #table::dsl;
+            use diesel::RunQueryDsl;
+
+            match table
+                .filter(dsl::id.eq(obj_id))
+                .get_result(conn)
+            {
+                Ok(doc) => Ok(doc),
+                Err(DieselError::NotFound) => Err(#errors_name ::NotFound { obj_id }.into()),
```

Comment:
I don't think that NotFound should always be treated as an error, in particular within an ORM where object inexistence is often checked. Using that if I want to check whether a Document exists, I need to explicitly match the Error variant `retrieve` returns. 

I think it would be more convenient to provide two sets of functions:
* `retrieve{,_conn}(...) -> Result<Option<Self>> ` where not found is not an error
* `find{,conn}(...) -> Result<Self>` where not found is an error

That way, checking for inexistence is a one-liner:

```rust
let exists: bool = Document::retrieve(pool, 42)?.is_some();
```

and so is finding a document that should exist:

```rust
let doc = Document::find(db_pool, document_key.into_inner()).await?;
```